File: solution/python/payment_client_fixed.py
"""
SOLUTION: Fixed PaymentClient — Python
Fixes: timeout enforcement, retry logic, proper exception propagation.
"""
import asyncio
import logging
import random
import yaml
import os

logger = logging.getLogger("payment_client")

_config_path = os.path.join(os.path.dirname(__file__), "../../config.yaml")
with open(_config_path) as f:
    _config = yaml.safe_load(f)

# FIX 4: Read timeout and retry from config
TIMEOUT = _config["payment_service"]["timeout"]
RETRY_COUNT = 3  # FIX 5: Add retry logic


class PaymentClient:
    async def process_payment(self, user_id: int, amount: float) -> dict:
        last_error = None

        # FIX 5: Retry loop
        for attempt in range(1, RETRY_COUNT + 1):
            try:
                logger.info(
                    f"Payment attempt {attempt}/{RETRY_COUNT} for user_id={user_id} amount={amount}"
                )

                # FIX 4: Enforce timeout using asyncio.wait_for
                result = await asyncio.wait_for(
                    self._call_payment_gateway(user_id, amount),
                    timeout=TIMEOUT,
                )

                logger.info(f"Payment successful for user_id={user_id} after {attempt} attempt(s)")
                return {"status": "success", "transaction_id": f"TXN-{user_id}-{attempt}"}

            except asyncio.TimeoutError:
                last_error = f"Payment gateway timeout after {TIMEOUT}s"
                logger.warning(f"Payment timeout on attempt {attempt} for user_id={user_id}")
            except Exception as e:
                last_error = str(e)
                logger.warning(
                    f"Payment error on attempt {attempt} for user_id={user_id}: {e}"
                )

            if attempt < RETRY_COUNT:
                await asyncio.sleep(0.5)

        logger.error(
            f"Payment failed after {RETRY_COUNT} attempts for user_id={user_id}: {last_error}"
        )
        # FIX 6: Return error status instead of silently failing
        return {"status": "error", "reason": last_error}

    async def _call_payment_gateway(self, user_id: int, amount: float) -> dict:
        """Simulates external payment gateway call."""
        await asyncio.sleep(2)
        if random.choice([True, False]):
            raise Exception("Payment gateway error: upstream timeout")
        return {"status": "success"}

File: solution/java/PaymentClientFixed.java
package com.shopsphere.checkout.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import java.util.Random;
import java.util.concurrent.CompletableFuture;

/**
 * SOLUTION: Fixed PaymentClient — Java
 * Fixes: retry logic, proper return type, exception propagation.
 */
@Component
public class PaymentClientFixed {

    private static final Logger logger = LoggerFactory.getLogger(PaymentClientFixed.class);
    private static final int MAX_RETRIES = 3;
    private final Random random = new Random();

    // FIX 2 + FIX 5: Return CompletableFuture so caller can await and check result
    @Async
    public CompletableFuture<String> processPaymentAsync(int userId, double amount) {
        String lastError = null;

        // FIX 5: Retry loop
        for (int attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            try {
                logger.info("Payment attempt {}/{} for userId={} amount={}", attempt, MAX_RETRIES, userId, amount);

                // FIX 4: Timeout would be enforced by the caller using CompletableFuture.get(timeout)
                Thread.sleep(2000);

                if (random.nextBoolean()) {
                    throw new RuntimeException("Payment gateway error: connection refused");
                }

                logger.info("Payment successful for userId={} after {} attempt(s)", userId, attempt);
                return CompletableFuture.completedFuture("success");

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                logger.error("Payment interrupted for userId={}", userId, e);
                return CompletableFuture.completedFuture("interrupted");

            } catch (Exception e) {
                lastError = e.getMessage();
                logger.warn("Payment attempt {} failed for userId={}: {}", attempt, userId, e.getMessage());

                if (attempt < MAX_RETRIES) {
                    try {
                        Thread.sleep(500); // backoff
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                    }
                }
            }
        }

        // FIX 6: Return failure status instead of silently swallowing
        logger.error("Payment failed after {} attempts for userId={}: {}", MAX_RETRIES, userId, lastError);
        return CompletableFuture.completedFuture("failed");
    }
}
